### PR TITLE
Add remote, parent & child configuration features

### LINKED ISSUE #1352: Enhancement: Allow inclusion of multiple configuration files
I'd like to propose adding a mechanism for including modular configuration files. Also requires the ability to name them other than .swiftlint.yml

i.e.:

```
disabled_rules:
  - trailing_whitespace
  - trailing_newline

[...stuff...]

include_subconfig:
  - customconfig1.yml                # some modular configuration (i.e. custom rules)
  - customconfig2.yml                # another modular configuration
```

I'm not sure the best way raise an enhancement request so opened an issue to gauge interest.

### PR DESCRIPTION
**Pull request by fredpi**
This PR closes #1352 and #2729.

It's basically the same as #2824, but I squashed all commits to keep it more comprehensible.

## Change Overview

- Using multiple configurations is now documented extensively in the README
- You can now specify `child_config` / `parent_config` configurations. This even works recursively, as long as there are no cycles and no ambiguities.
- Nested configurations got a speed bump (via caching) and of course now ignore configs that were already used to build the primary config (using `child_config` / `parent_config` references)
- You can pass multiple configurations via the command line that are interpreted as a hierarchy
- You can specify a remote `child_config` / `parent_config`! This is just easy as dropping in a url that starts with `http://`or `https://` where normally there would be the path to a local file. Remote configs get cached (in a manner that git doesn't see it). There are timeouts to loading remote configs, that can also be customized using `remote_timeout` / `remote_timeout_if_cached`
- Rule Configurations now get merged properly (Previously, the rule configurations from the parent configuration were always used). This change results in the following behavior:
   - When a child config doesn't configure a rule, yet the parent config does, the parent config will be used
    - When a child config configures a rule, and the parent doesn't, the child config will be used
    - When a child config configures a rule, and the parent does, too, the child config will be used
- Warnings about duplicate and invalid identifiers now appear as actual warnings in Xcode
- Internally, the whole `Configuration` clutter has been cleaned up in a way that I hope everyone is happy with. I strongly felt like the current merging implementation that relied on a `rules` array and a `rulesMode` didn't express the very foundation of what a configuration is: A mode that is applied to an array of all available rules with configurations, resulting in an array of rules that should be linted. To express this very relation, I created the new `RulesWrapper` type. While offering a `resultingRules` variable that is lazily computed, it still retains what caused these rules and therefore eases understanding of the merging algorithm.

#### Sidenote: Fixed Warnings in Moya

SwiftLintBot will probably complain that, with this new implementation, 3 warnings in Moya are now fixed. In fact, previously those have been false positives, as in the nested `.swiftlint.yml` covering these files there's the following configuration for the `file_length` rule:

```
file_length:
  warning: 1000
```

As the parent rule configuration was used for merging and there was no smart mechanism to decide which configuration to choose, the configuration originating from the nested configuration was ignored. With this PR, that's now fixed.

### CODE REVIEW

**Comment by fredpi** [comment]
@jpsim I added the missing tests (didn't come up with a proper idea how to test remote configs though), so this is now finally ready for review 🎉

**Comment by fredpi** [comment]
I just rebased to fix merge conflicts

**Comment by fredpi** [comment]
I just rebased to fix conflicts.

**Comment by fredpi** [comment]
I just rebased to fix conflicts.

**Comment by mokagio** [comment]
Is there anything I can do to help this get merged and shipped?

**Comment by fredpi** [comment]
@mokagio Thanks for asking and offering your help! 👍

From my side, this PR is ready and only needs a review. But as it changes a lot of things quite significantly, such a review

- will take time
- should preferably be done by someone with in-depth-knowledge about this project (like @jpsim or @marcelofabri)

But as this is a big PR and especially with the current situation, nobody can blame them for not having reviewed this yet or expect them to review this right now.

So, what you can do, is what you already did: Test this PR in real projects (and post your results in this thread), so that we can find bugs or other issues that may still exist. That's really helpful and gives confidence that this PR actually works in real environments! 🚀

Also, if you have the time, you may also do a review! I just think that @jpsim or @marcelofabri or another experienced SwiftLint contributor should approve this PR, but that doesn't say that other opinions aren't valuable!

**Comment by fredpi** [comment]
I just rebased to fix merge conflicts.

**Review by jpsim — APPROVED** [review]
I'm sorry @fredpi, you've continuously done awesome work on this feature and I've repeatedly failed to give prompt feedback or reviews. Given that you have write access to the repo, I encourage you to merge this yourself as soon as you've addressed the merge conflicts and you're comfortable with the changes.

I'd like to hold a higher standard for actively participating and helping contributors such as yourself, but life gets in the way.

File: README.md
Comment (jpsim) [inline README.md:498]: Nit extra space:

```suggestion
You can specify a `child_config` and / or a `parent_config` reference within a configuration file.
```

File: README.md
Comment (jpsim) [inline README.md:525]: Maybe just "Remote"?

```suggestion
### Child / Parent Configs (Remote)
```

File: README.md
Comment (jpsim) [inline README.md:540]: Just to make sure, if there are no config references starting with `http(s)://`, no network calls are made, right?

```suggestion
Every time you run SwiftLint and have an Internet connection, SwiftLint tries to get a new version of
```

File: README.md
Comment (jpsim) [inline README.md:551]:
```suggestion
Instead of just providing one configuration file when running SwiftLint via the command line,
```

File: README.md
Comment (jpsim) [inline README.md:572]: This is indented by a single space:

```suggestion
`.swiftlint.yml` files are only considered as a nested configuration if they have not been
used to build the main configuration already (e. g. by having been referenced via something
like `child_config: Folder/.swiftlint.yml`). Also, `parent_config` / `child_config`
specifications of nested configurations are getting ignored because there's no sense to that.
```

File: Source/SwiftLintFramework/Extensions/Configuration+Cache.swift
Comment (jpsim) [inline Source/SwiftLintFramework/Extensions/Configuration+Cache.swift:94]: I try to never indent compiler conditionals like these. Can you please revert?

File: Source/SwiftLintFramework/Helpers/Reachability.swift
Comment (jpsim) [inline Source/SwiftLintFramework/Helpers/Reachability.swift:9]: Maybe this instead?

```suggestion
    internal static func isConnectedToNetwork() -> ConnectivityStatus {
```

where

```swift
enum ConnectivityStatus { case connect, disconnected, unknown }
```

File: Source/SwiftLintFramework/Helpers/Reachability.swift
Comment (jpsim) [inline Source/SwiftLintFramework/Helpers/Reachability.swift:22]: Shouldn't this be `nil` or `.unknown`?

File: Source/SwiftLintFramework/Helpers/Reachability.swift
Comment (jpsim) [inline Source/SwiftLintFramework/Helpers/Reachability.swift:11]: Shouldn't this be `nil` or `.unknown`?

**Comment by fredpi** [comment]
Thank you very much @jpsim and please don't feel sorry – this is a huge project and nobody can expect you to spend even more of your free time than you already do ❤️

The rebase I did today was quite complex and I haven't yet adapted my code to the changes e. g. introduced by https://github.com/realm/SwiftLint/pull/3379. I will try to do this some time next week, then work through your feedback, test everything again, and then possibly merge this PR.

File: README.md
Comment (fredpi) [reply README.md:556]: Nope, no network calls are made ;)

**Comment by fredpi** [comment]
Quick update: I finished the rebase and implemented the review feedback, but noticed that I still have to take a closer look on the `rootPath` / `configurationPath` semantics again. I hope to find some time for this within the next days.

**Comment by sethfri** [comment]
@fredpi This is incredible! I'd love to look this over and give feedback if you'd like once you're done finishing up what you mentioned in your last comment.

That said, this has been outstanding for a long time, so I don't blame you if you just want to get it merged

**Comment by fredpi** [comment]
@sethfri While I didn't test everything myself yet, I don't see any major issues anymore, so I'd love it if you would try out the PR and report any issues you see! 👍 Also, it would be great if you could have a look at the new behavior specification for multiple configurations in the README and tell me whether you agree with these specifications.

Of course, before this will get merged, I will also test everything myself again.

**Comment by fredpi** [comment]
@mokagio It would be great if you could also report whether this PR is still working for you ;)

**Comment by mokagio** [comment]
@fredpi hey there! You're awesome for relentlessly working on this ❤️ 

I updated our experiment repo to use `ed50938bb955f224bf9146eb0b89e1368e4aca67`. It builds and runs fine, both locally and on CI. 👍 

The only thing I noticed is that it's not respecting the `excluded` setting 🤔 .

As you can see in [this CI build](https://github.com/wordpress-mobile/WordPress-iOS-Shared/runs/1421667168?check_suite_focus=true#step:7) we're getting violation warnings from files in the `Pods/` folder, but that's [excluded in the remote config being used].(https://github.com/Automattic/swiftlint-config/blob/59fc2405960928f7fddb6a5e0fb77b6ce8d6f9c5/.swiftlint.yml#L2-L3)

_Screenshot included because as I discovered GitHub Actions doesn't retain the logs indefinitely (for free repos, at least). The first line is from a file in `Pods/` and unexpected, the second from one of the sources in the project and expected._

image

**Comment by fredpi** [comment]
@mokagio Thanks for the feedback! 😃

I could reconstruct and understand the bug with the `excluded` paths being ignored (+ write a test for it). With the latest commit (`a37d4487f071463b619a6ff7367edcb50daeb5ca`) the bug should be fixed now. Would it be possible for you to test the latest version again and confirm whether the issue is now gone?

**Comment by mokagio** [comment]
@fredpi thank you for your quick work! I updated to `a37d4487f071463b619a6ff7367edcb50daeb5ca` and it all looks good no. See [this CI run](https://github.com/wordpress-mobile/WordPress-iOS-Shared/runs/1428054238?)

👍 🙌 👏

**Comment by fredpi** [comment]
I have added some more tests, tested the PR myself again and feel quite confident now that everything is working properly. So, if no new concerns get raised until then, I'm going to **merge this on Sunday**.

@sethfri If you still want to take a look at this and don't have the time to do it until Sunday, please tell me – I would then delay the merging.

**Comment by mokagio** [comment]
I'm so happy this is in! ❤️

**Pull request merged**
